lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,422 @@
+  double memoryConsumption = config.MBlock * config.NBlock +
+                             config.NBlock * config.KBlock +
+                             config.MBlock * config.KBlock;
+  double computationIntensity = FLOPS / memoryConsumption;
+  if (memoryConsumption * (dtypeSize / bitPerByte) > L2Cache * fullLoadRatio) {
+    computationIntensity /= outOfCachePenalty;
+  }
+  return 1 / computationIntensity;
+}
+
+using CostModelFn =
+    std::function<float(linalg::LinalgOp &linalgOp, ArrayRef<uint32_t> shape,
+                        MatmulConfig cfg, SystemDesc &sysDesc)>;
+
+// filter the config by the cost model
+std::vector<MatmulConfig>
+filterConfigByCostModel(std::vector<MatmulConfig> configs,
+                        linalg::LinalgOp &linalgOp, ArrayRef<uint32_t> shape,
+                        SystemDesc &sysDesc, const CostModelFn &costModel,
+                        float eliminationRatio = 0.5, float threshold = -1) {
+  std::vector<MatmulConfig> result;
+  std::vector<float> costs;
+  std::vector<size_t> idx;
+  for (auto [i, config] : llvm::enumerate(configs)) {
+    costs.push_back(costModel(linalgOp, shape, config, sysDesc));
+    idx.push_back(i);
+  }
+  std::stable_sort(idx.begin(), idx.end(), [&costs](size_t i1, size_t i2) {
+    return costs[i1] < costs[i2];
+  });
+  double thresholdCost =
+      costs[idx[(size_t)(eliminationRatio * configs.size())]];
+  thresholdCost =
+      threshold < thresholdCost && threshold > 0 ? threshold : thresholdCost;
+  for (size_t i = 0; i < configs.size(); i++) {
+    if (costs[idx[i]] <= thresholdCost) {
+      result.push_back(configs[idx[i]]);
+    }
+  }
+  LLVM_DEBUG(llvm::dbgs() << "thresholdCost is: " << thresholdCost
+                          << "\nbest with cost: " << costs[idx[0]] << "\n"
+                          << configs[idx[0]] << "\n worst with cost: "
+                          << costs[idx[configs.size() - 1]] << "\n"
+                          << configs[idx[configs.size() - 1]] << "\n");
+  return !result.empty() ? result : configs;
+}
+
+// prepare the config candidates
+std::vector<MatmulConfig>
+prepareConfigCandidates(Operation *root, SystemDesc &sysDesc,
+                        ArrayRef<uint32_t> shape,
+                        ArrayRef<uint32_t> givenInnermostBlock) {
+  std::vector<MatmulConfig> configs;
+  uint32_t threads = sysDesc.getNumThreads();
+  std::vector<uint32_t> MThreadsCandidates =
+      getCandidate((uint32_t)threads, 1U);
+  std::vector<uint32_t> NThreadsCandidates =
+      getCandidate((uint32_t)threads, 1U);
+  std::vector<uint32_t> KThreadsCandidates =
+      getCandidate((uint32_t)threads, 1U);
+  uint32_t noSmallBlockNeedThreshold = 8 * 8U;
+  std::vector<uint32_t> MBlockCandidates = getCandidate(
+      (uint32_t)shape[0], shape[0] > noSmallBlockNeedThreshold ? 8U : 1U,
+      (uint32_t)shape[0]);
+  std::vector<uint32_t> NBlockCandidates =
+      getCandidate((uint32_t)shape[1],
```

Comment:
Check the size of `shape` before indexing it.

Reply:
Add a check here

Reply:
forgot?